core/src/main/scala/scala/pickling/generator/CaseClassPickling.scala

```diff
@@ -0,0 +1,161 @@
+package scala.pickling
+package generator
+
+
+/** this algorithm inspects symbols to determine if we have a scala case class, and generates the
+  * pickling code for it.
+  *
+  * This ONLY handles case-class types, it will not handle ADTS, but can generate code for non-final case classes.
+  *
+  */
+class CaseClassPickling(val allowReflection: Boolean) extends PicklingAlgorithm {
+  case class FieldInfo(name: String, sym: IrMethod)
+  case class CaseClassInfo(constructor: IrConstructor, fields: Seq[FieldInfo])
+
+
+  // TODO - This helper method should be available elsewhere.
+  def allVars(cls: IrClass): Seq[IrMethod] = {
+    ( cls.methods.filter(_.isParamAccessor) ++
+      IrSymbol.allDeclaredMethodIncludingSubclasses(cls).filter(x => x.isVar || x.isVal)
+    ).groupBy(_.methodName).map(_._2.head).toList.filterNot(_.isMarkedTransient)
+  }
+  private def checkConstructorImpl(tpe: IrClass, logger: AlgorithmLogger): AlgorithmResult = {
+    if(tpe.isCaseClass) {
+      tpe.primaryConstructor match {
+        case Some(c) if c.isPublic =>
+          val names = c.parameterNames.flatten.toSet
+          val vars = allVars(tpe)
+          val standAloneVars = vars.filter { m =>
+            m.isVar && !(names contains m.methodName)
+          }
+          // TODO - Allow us to DISABLE serialziing these vars.
```

Comment:
Haha

Maybe it's time to correct the typos in public interface methods, too? ;-)